pkg/cmdline/cmd.go

```
@@ -0,0 +1,124 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package cmdline
+
+import (
+	"fmt"
+
+	"github.com/spf13/cobra"
+)
+
+// CommandManager holds root command or first parent
+// and can stores group of command
+type CommandManager struct {
+	rootCmd   *cobra.Command
+	groupCmds map[string][]*cobra.Command
+	errPool   []error
+	fm        *flagManager
+}
+
+// NewCommandManager instantiates a CommandManager
+func NewCommandManager(rootCmd *cobra.Command) *CommandManager {
+	if rootCmd == nil {
+		panic("nil root command passed")
+	}
+	cm := &CommandManager{
+		rootCmd:   rootCmd,
+		groupCmds: make(map[string][]*cobra.Command),
+		errPool:   make([]error, 0),
+		fm:        newFlagManager(),
+	}
+	return cm
+}
+
+func (m *CommandManager) pushError(f string, a ...interface{}) {
+	m.errPool = append(m.errPool, fmt.Errorf(f, a...))
+}
+
+// GetError returns the error pool
+func (m *CommandManager) GetError() []error {
+	return m.errPool
+}
+
+// RegisterCmd registers a child command for the root command
+func (m *CommandManager) RegisterCmd(cmd *cobra.Command, interspersed bool) {
+	// panic here because it's a misuse of API and generally from
+	// global context or init() functions
+	if cmd == nil {
+		panic("nil command passed")
```

please don't panic, please return errors.

---

It's really intentional and decided to add `panic` because there is no interaction with user input here, those conditions are not supposed to be reached when used correctly, so if developer does something wrong, code just panic with an explicit message instead of simply ignoring this check like most lib

---

Not sure about this. In the specific case you describe, without this you would still get a panic from the next line (cmd.SetFlagErrorFunc) because that function tries to access a field in cmd (in Go it's valid to call a method with a nil receiver, so calling the function does not panic). From that point of view, the only thing this is doing is changing the message that gets displayed. You will still have to walk up the stack to figure out why cmd is nil.

I guess I'm saying this specific use is OK.

---

Those panic are not supposed to happen, if it occurs that means CLI manager wasn't correctly used and triggered by developer mistake